clippy_lints/src/question_mark.rs

```diff
@@ -531,6 +589,41 @@ impl<'tcx> LateLintPass<'tcx> for QuestionMark {
         self.try_block_depth_stack.push(0);
     }
 
+    // Defining check_fn instead of using check_body to avoid accidentally triggering on
+    // const expressions, not sure if this is necessary
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        _: &'tcx FnDecl<'tcx>,
+        body: &'tcx Body<'tcx>,
+        _: Span,
+        _: LocalDefId,
+    ) {
+        let return_expr;
```

Comment:
Can we make it more compact something like this

  ```
 let return_expr = match body.value.kind {
        ExprKind::Block(block, _) => block.expr.or_else(|| {
            block.stmts.iter().last().and_then(|stmt| {
                if let StmtKind::Semi(expr) = &stmt.kind {
                    if let ExprKind::Ret(Some(ret)) = &expr.kind {
                        return Some(ret);
                    }
                }
                None
            })
        }),
        _ => None,
    };

```
    if let Some(expr) = return_expr {
        check_if_let_some_as_return_val(cx, expr);
    }

Reply:
Thanks for the suggestion, this is much better